### DISCUSSION TITLE
Set response header in a custom rust plugin using ControlFlow::Continue

### DISCUSSION THREAD

**Discussion by Meemaw** [apollographql/router#3451] [General]
Hey,

Is it possible to set a response header in a custom rust plugin when using `ControlFlow::Continue`? Its obviously possible to do so when using `ControlFlow::Break` when a new response is built, but unsure about `Continue`.

**Comment by garypen** [comment]
You could set the header and it will be propagated through to the subgraphs. However, I think that you are looking for a header to be propagated back to the client.

I'd implement this by setting a key in context when you know you are continuing and then check if the key is present in one of the response services (router or supergraph would be good choices). If the context key is present, then set the header there from the context value. You could do this in rhai or in your custom plugin.

**Comment by Meemaw** [comment]
Hey @garypen thanks for the help. Came up with something like this, does this looks rougly okay?

```rust
    .map_response(|mut response: supergraph::Response| {
        let context_value: Result<Option<String>, BoxError> =
            response.context.get(header);

        if let Ok(Some(remaining_tokens)) = context_value {
            let header = HeaderValue::from_str(&remaining_tokens);
            if let Ok(header_value) = header {
                response
                    .response
                    .headers_mut()
                    .append(header, header_value);
            }
        }
        response
    })
 ```
 
 Actually, while at it, I have a question on how to test that this header is present in the response. Testing the plugin using `TestHarness` + `oneshot`, but the `graphql::Response` doesn't seem to expose headers?

**Reply by bnjjj** [reply]
For the test, if you're using something like this:

```rust
let service_stack = YourPlugin::new(PluginInit::try_builder().config(config).supergraph_sdl(Default::default()).build().unwrap())
    .await
    .unwrap()
    .supergraph_service(mock_service.boxed());
let res = service_stack
    .oneshot(request)
    .await
    .unwrap();

// Access headers via res.response.headers();
```
